File: 0139-Word Break/0139-Word Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        if (n == 0)
        {
            return true;
        }
        
        vector<bool> dp(1 + n);
        dp[0] = true;
        int maxLen = 0;
        unordered_set<string> wordSet;
        for (string w : wordDict)
        {
            maxLen = max(maxLen, (int)(w.size()));
            wordSet.insert(w);
        }
        
        for (int j = 1; j <= n; ++j)
        {
            for (int i = j - 1; i >= j - maxLen && i >= 0; --i)
            {
                if (wordSet.find(s.substr(i, j - i)) != wordSet.end() && dp[i])
                {
                    dp[j] = true;
                    break;
                }
            }
        }
        
        return dp[n];
    }
};

File: 0101-0200/0139-Word Break/0139-Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set()
        maxLen = 0
        for word in wordDict:
            wordSet.add(word)
            maxLen = max(maxLen, len(word))
        dp = [False] * (1 + len(s))
        dp[0] = True
        for j in range(1, len(dp)):
            for i in range(j - 1, -1, -1):
                if j - i > maxLen:
                    break
                if s[i:j] in wordSet and dp[i]:
                    dp[j] = True
                    break
        return dp[-1]
